docs/DevGuide/Brigand.md

```diff
@@ -0,0 +1,2015 @@
+header:
+\snippet Test_TMPLDocumentation.cpp include
+
+\par
+Examples in this document use the following declarations and definitions:
+\snippet Test_TMPLDocumentation.cpp example_declarations
+
+
+\section Metafunctions
+
+\par
+A metafunction is an analog of a familiar C++ function that is coded in the C++
+type system.  It turns out that, using metafunction programming, it is possible
+to perform arbitrary computations at compile time.
+
+\par
+There are multiple ways to encode a calculation in the type system.  When using
+Brigand, the relevant ones are eager and lazy metafunctions.
+
+
+\subsection lazy Eager and lazy metafunctions
+
+\par
+Metafunctions commonly appear in two forms: eager and lazy.  Lazy metafunctions
+are templated structs (or templated aliases to structs) with a `type` member
+alias that indicates the result:
+\snippet Test_TMPLDocumentation.cpp metafunctions:lazy
+The type traits in the standard library, such as std::is_same, are lazy
+metafunctions.
+
+\par
+Eager metafunctions are aliases to their result types.  As a trivial case,
+struct templates can be viewed as eager metafunctions returning themselves.  An
+eager version of the previous example could be implemented as:
+\snippet Test_TMPLDocumentation.cpp metafunctions:eager
+The standard library provides eager versions of some of its metafunctions
+(generally those that modify a type, rather than predicates) using an `_t`
+suffix.  When both versions are provided, it is often convenient (and less
+error prone!) to define the eager version in terms of the lazy version:
+\snippet Test_TMPLDocumentation.cpp metafunctions:eager_from_lazy
+And the two definitions agree:
+\snippet Test_TMPLDocumentation.cpp metafunctions:agreement
+
+\note
+The standard library also provides versions of many of its type traits with an
+`_v` suffix.  These evaluate to compile-time *values*, rather than types.  They
+can be useful for metaprogramming, but are not the types of metafunctions being
+discussed here.
+
+\par
+Eager metafunctions are usually more convenient to use, so what is the point of
+additionally creating lazy ones?  The answer is that lazy metafunctions can be
+used as compile-time functors.  As a simple example, we can write an (eager)
+metafunction that calls an arbitrary lazy metafunction twice
+\snippet Test_TMPLDocumentation.cpp metafunctions:call_lazy_metafunction
+and get the expected output:
+\snippet Test_TMPLDocumentation.cpp metafunctions:call_lazy_metafunction_assert
+But it fails if you try to call an arbitrary *eager* metafunction
+twice in the same way, because the function is evaluated too early,
+resulting in the `List1` metafunction being acted upon instead of
+`eager_add_list`:
+\snippet Test_TMPLDocumentation.cpp metafunctions:call_eager_metafunction
+\snippet Test_TMPLDocumentation.cpp metafunctions:call_eager_metafunction_assert
+(In this simple case we could have used a template template parameter to pass
```

Comment:
Thought `template template` was a typo the first time I saw it. Maybe add a comment that this isn't a typo and refers to `template<template<...>>`

Reply:
I added a link to cppreference.